### ISSUE TITLE
Add process for obtaining cluster assignments

### ISSUE THREAD

**Issue by allyhawkins** [AlexsLemonade/scpca-nf#404]
We would like to provide cluster assignments in the SCE objects output from the workflow. These will ultimately be compared to the cell type annotations and shown in the QC report, but we will first need to add them to the objects. Some things we will need to figure out: 

I think we want to do this within the main workflow prior to cell typing. For now that means adding a step after the normalization and dim reduction step that adds in clusters. They would then be returned in the processed SCE object which is currently used as input to the cell type workflow.

**Comment by sjspielman** [comment]
From our discussion, we'll use `leiden` as default for clustering since that's where the field is actively headed, but we'll keep it as a parameter in the workflow for flexibility.

From an even earlier discussion, we'll do this in a separate script to live in (conceptually..) `bin/script-that-does-clustering.R` and add a process into`modules/process-that-calls-clustering-script.nf` that will be called between `post_process_sce` and `sce_qc_report` in the main workflow - 
https://github.com/AlexsLemonade/scpca-nf/blob/9efd7734deb0d807ecbc06ecce4ac99a3a9a0b33/main.nf#L210-L214

**Comment by allyhawkins** [comment]
That is a lot of clusters for Leiden! Did you look at them on a UMAP? Does it make sense that there are that many or do you see a lot of mixing between clusters?

**Comment by sjspielman** [comment]
Well, this certainly did change the results!! Though, possibly not as hoped....

- Number of iterations doesn't have much of an effect
- If I specify use "jaccard" with pretty much any k, I get 4864 clusters (guess how many cells are in this SCE? that's right, 4864). The default type gives # clusters in the 20s, as observed.
- There is a resolution parameter I can supply to `cluster.args`, which I played around with a bit at `k=15`:
```
res_params <-  seq(0.001, 0.5, 0.025)
purrr::walk(
  res_params, 
  \(x) {
    jaccard_clusters <- bluster::clusterRows(
      pca_matrix, 
      bluster::NNGraphParam(
        k = 15, 
        cluster.fun = "leiden",
        type = "jaccard",
        cluster.args = list(
          resolution = x
      ))) |> levels() |> length()
      default_clusters <- bluster::clusterRows(
        pca_matrix, 
        bluster::NNGraphParam(
          k = 15, 
          cluster.fun = "leiden",
          cluster.args = list(
            resolution = x
        ))) |> levels() |> length()
      print(glue::glue("{x}: {jaccard_clusters} | {default_clusters}"))
    }
)
```

And results:
```
0.001: 5 | 1
0.026: 135 | 3
0.051: 378 | 3
0.076: 678 | 3
0.101: 978 | 5
0.126: 1454 | 6
0.151: 1940 | 10
0.176: 2183 | 8
0.201: 2711 | 9
0.226: 2884 | 9
0.251: 3350 | 8
0.276: 3462 | 8
0.301: 3864 | 9
0.326: 3935 | 11
0.351: 4221 | 10
0.376: 4279 | 10
0.401: 4449 | 13
0.426: 4473 | 13
0.451: 4488 | 14
0.476: 4618 | 13
```

My conclusion: leiden plays _poorly_ with jaccard.

**Comment by jashapiro** [comment]
> My conclusion: leiden plays _poorly_ with jaccard.

It's maybe not quite that bad... According to https://igraph.org/r/doc/cluster_leiden.html the optimal clustering from a weighted graph is when resolution = 1/2m, where m is the number of edges, so we should be looking at pretty small resolutions here, I think. (Now, finding the number of edges may be a bit more trouble than we want here, but it is almost always going to be more than the number of cells, which puts it below into the lowest value you have here).

**Comment by jashapiro** [comment]
All that said, it may be that going with `type="jaccard"` and `cluster.fun="louvain"` is the right move for now... The extra parameters/bad defaults for Leiden clustering may need some time to work out.

**Comment by sjspielman** [comment]
Closed by #405
